### PR TITLE
Use string interning in statskeeper for the endpoint path

### PR DESCRIPTION
**Pull request by val06**
### What does this PR do?

Replaced `Path.Content` field to use string interning, that way we can save some memory allocations when similar endpoints are hit again and again

### Motivation

Reduce memory usage and optimize temporary allocations

### Additional Notes

Load tests: 
_(all tests are re-using the connections and don't create new connections, which should reduce the number of total entries in  StatsKeeper's map)_

to run:

`inv -e usm-load-test.http-scenario --env val-string-intern --api-key <api_key> --compare-docker-repo 601427279990.dkr.ecr.us-east-1.amazonaws.com/usm-agent/valeri.pliskin --compare-agent string_intern --scenarios medium_volume_client --base-docker-repo 601427279990.dkr.ecr.us-east-1.amazonaws.com/usm-agent/valeri.pliskin --base-agent no_string_intern`

performance dashboards:

- medium load [dashboard](https://dddev.datadoghq.com/dashboard/gqu-2wz-asy/usm-performance-evaluation-20?refresh_mode=paused&tpl_var_base_agent-env%5B0%5D=val-string-intern-base&tpl_var_client-service%5B0%5D=http-k6-client&tpl_var_compare_agent-env%5B0%5D=val-string-intern-new&tpl_var_kube_cluster_name%5B0%5D=usm-test-env-sso&tpl_var_server-service%5B0%5D=golang-httpbin&from_ts=1695026623170&to_ts=1695030342314&live=false)
      - [profiler](https://dddev.datadoghq.com/profiling/comparison?query=agent-env%3Aval-string-intern-new%20&compare_end_A=1695030300000&compare_end_B=1695030300000&compare_facet=log_prof_go_lifetime_heap_bytes&compare_query_A=agent-env%3Aval-string-intern-base&compare_query_B=agent-env%3Aval-string-intern-new&compare_start_A=1695026580000&compare_start_B=1695026580000&compareValuesMode=absolute&comparisonViz=flamegraph&my_code=enabled&op_filter=focus_on%28function%3A%22%28%2AStatKeeper%29.Process%20%28statkeeper.go%29%22%29&profile_type=heap-live-size&viz=stream&start=1694678791000&end=1694679391000&paused=true)

- production load [dashboard](https://dddev.datadoghq.com/dashboard/gqu-2wz-asy/usm-performance-evaluation-20?refresh_mode=paused&tpl_var_base_agent-env%5B0%5D=val-string-intern-base&tpl_var_client-service%5B0%5D=http-k6-client&tpl_var_compare_agent-env%5B0%5D=val-string-intern-new&tpl_var_kube_cluster_name%5B0%5D=%2A&tpl_var_server-service%5B0%5D=golang-httpbin&from_ts=1694692800000&to_ts=1694706000000&live=false)

#### Benchmarks:

paths: 10000
unique paths: 50
unique conn tuples = 30

**BASELINE (no intern)**

```
goos: linux
goarch: amd64
pkg: github.com/DataDog/datadog-agent/pkg/network/protocols/http
cpu: 11th Gen Intel(R) Core(TM) i9-11950H @ 2.60GHz
BenchmarkProcessHTTPTransactions
BenchmarkProcessHTTPTransactions-16    	      10	   5087173 ns/op	  148880 B/op	    1179 allocs/op
BenchmarkProcessHTTPTransactions-16    	      10	   4352453 ns/op	  148518 B/op	    1178 allocs/op
BenchmarkProcessHTTPTransactions-16    	      10	   4424486 ns/op	  148540 B/op	    1178 allocs/op
BenchmarkProcessHTTPTransactions-16    	      10	   4163176 ns/op	  148409 B/op	    1178 allocs/op
BenchmarkProcessHTTPTransactions-16    	      10	   5086082 ns/op	  148729 B/op	    1179 allocs/op
```

**String Interning**

```
goos: linux
goarch: amd64
pkg: github.com/DataDog/datadog-agent/pkg/network/protocols/http
cpu: 11th Gen Intel(R) Core(TM) i9-11950H @ 2.60GHz
BenchmarkProcessHTTPTransactions
BenchmarkProcessHTTPTransactions-16    	      10	   4842102 ns/op	  147274 B/op	    1174 allocs/op
BenchmarkProcessHTTPTransactions-16    	      10	   4388956 ns/op	  146602 B/op	    1172 allocs/op
BenchmarkProcessHTTPTransactions-16    	      10	   4215041 ns/op	  146823 B/op	    1173 allocs/op
BenchmarkProcessHTTPTransactions-16    	      10	   4332209 ns/op	  146663 B/op	    1172 allocs/op
BenchmarkProcessHTTPTransactions-16    	      10	   4996685 ns/op	  146775 B/op	    1172 allocs/op
```

------------------------------------

paths: 10000
unique paths: 50
unique conn tuples  = 1000

**BASELINE (no intern)**

```
goos: linux
goarch: amd64
pkg: github.com/DataDog/datadog-agent/pkg/network/protocols/http
cpu: 11th Gen Intel(R) Core(TM) i9-11950H @ 2.60GHz
BenchmarkProcessHTTPTransactions
BenchmarkProcessHTTPTransactions-16    	      10	   3351261 ns/op	  478958 B/op	    5309 allocs/op
BenchmarkProcessHTTPTransactions-16    	      10	   3734514 ns/op	  479131 B/op	    5309 allocs/op
BenchmarkProcessHTTPTransactions-16    	      10	   3479820 ns/op	  478724 B/op	    5309 allocs/op
BenchmarkProcessHTTPTransactions-16    	      10	   3833522 ns/op	  478708 B/op	    5308 allocs/op
BenchmarkProcessHTTPTransactions-16    	      10	   3441786 ns/op	  478596 B/op	    5308 allocs/op
```

**String Interning**

```
goos: linux
goarch: amd64
pkg: github.com/DataDog/datadog-agent/pkg/network/protocols/http
cpu: 11th Gen Intel(R) Core(TM) i9-11950H @ 2.60GHz
BenchmarkProcessHTTPTransactions
BenchmarkProcessHTTPTransactions-16    	      10	   4117142 ns/op	  473944 B/op	    5304 allocs/op
BenchmarkProcessHTTPTransactions-16    	      10	   3508454 ns/op	  473823 B/op	    5303 allocs/op
BenchmarkProcessHTTPTransactions-16    	      10	   3387072 ns/op	  473765 B/op	    5303 allocs/op
BenchmarkProcessHTTPTransactions-16    	      10	   3685999 ns/op	  474501 B/op	    5305 allocs/op
BenchmarkProcessHTTPTransactions-16    	      10	   3466970 ns/op	  474069 B/op	    5304 allocs/op
```
------------------------------------

paths: 10000
unique path: 50
always new port 

**BASELINE (no intern)**

```
goos: linux
goarch: amd64
pkg: github.com/DataDog/datadog-agent/pkg/network/protocols/http
cpu: 11th Gen Intel(R) Core(TM) i9-11950H @ 2.60GHz
BenchmarkProcessHTTPTransactions
BenchmarkProcessHTTPTransactions-16    	      10	   9349743 ns/op	 2460779 B/op	   38023 allocs/op
BenchmarkProcessHTTPTransactions-16    	      10	   8121694 ns/op	 2458891 B/op	   38020 allocs/op
BenchmarkProcessHTTPTransactions-16    	      10	   8077411 ns/op	 2461560 B/op	   38025 allocs/op
BenchmarkProcessHTTPTransactions-16    	      10	   7985048 ns/op	 2460225 B/op	   38023 allocs/op
BenchmarkProcessHTTPTransactions-16    	      10	   8391819 ns/op	 2460705 B/op	   38024 allocs/op
```

**String Interning**

```
goos: linux
goarch: amd64
pkg: github.com/DataDog/datadog-agent/pkg/network/protocols/http
cpu: 11th Gen Intel(R) Core(TM) i9-11950H @ 2.60GHz
BenchmarkProcessHTTPTransactions
BenchmarkProcessHTTPTransactions-16    	      10	   7681500 ns/op	 2430674 B/op	   38019 allocs/op
BenchmarkProcessHTTPTransactions-16    	      10	   8330516 ns/op	 2429461 B/op	   38017 allocs/op
BenchmarkProcessHTTPTransactions-16    	      10	   8590826 ns/op	 2430293 B/op	   38018 allocs/op
BenchmarkProcessHTTPTransactions-16    	      10	   7929292 ns/op	 2429698 B/op	   38017 allocs/op
BenchmarkProcessHTTPTransactions-16    	      10	   9083679 ns/op	 2429864 B/op	   38018 allocs/op
```

### Possible Drawbacks / Trade-offs

string interner pkg is a thread safe pkg and using a mutex for each access

### Describe how to test/QA your changes

usm-load-test env
use `BenchmarkProcessHTTPTransactions` benchmark test with different values to compare vs the base version (without string interning)

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/util/intern/string.go
Comment (guyarb) [inline pkg/util/intern/string.go:1]: seems irrelevant to the PR, I'd remove that (and it will remove ebpf-platform as a reviewer)

File: pkg/network/protocols/http/stats.go [hunk 1]
```diff
@@ -77,11 +81,23 @@ type Key struct {
...
+func NewKey(saddr, daddr util.Address, sport, dport uint16, path []byte, fullPath bool, method Method) Key {
 	return Key{
 		ConnectionKey: types.NewConnectionKey(saddr, daddr, sport, dport),
 		Path: Path{
-			Content:  path,
+			Content:  Interner.Get(path),
+			FullPath: fullPath,
+		},
+		Method: method,
+	}
```
Comment (guyarb) [inline pkg/network/protocols/http/stats.go:92]: return NewKeyWithConnection(types.NewConnectionKey(saddr, daddr, sport, dport), path, fullPath, method)

File: pkg/network/protocols/http/stats.go [hunk 2]
```diff
@@ -6,13 +6,17 @@
 package http
 
 import (
+	"github.com/DataDog/datadog-agent/pkg/util/intern"
```
Comment (guyarb) [inline pkg/network/protocols/http/stats.go:9]: please move to the correct location

**Review by guyarb — APPROVED** [review]
Small changes

File: pkg/util/intern/string.go
Comment (val06) [reply pkg/util/intern/string.go:1]: i thought i reverted those.. will do it

**Review by leeavital — APPROVED** [review]
LGTM for networks code

File: pkg/network/protocols/http/stats.go [hunk 3]
```diff
@@ -10,9 +10,13 @@ import (
 
 	"github.com/DataDog/datadog-agent/pkg/network/types"
 	"github.com/DataDog/datadog-agent/pkg/process/util"
+	"github.com/DataDog/datadog-agent/pkg/util/intern"
 	"github.com/DataDog/datadog-agent/pkg/util/log"
 )
 
+// Interner is used to intern strings to save memory allocations.
+var Interner = intern.NewStringInterner()
```
Comment (hmahmood) [inline pkg/network/protocols/http/stats.go:18]: Does this need to be exported?

**Review by hmahmood — APPROVED** [review]

File: pkg/network/protocols/http/stats.go (hunk 3 above)
Comment (val06) [reply pkg/network/protocols/http/stats.go:18]: yes, unfortunately it is used by some tests that are not part of the `http` pkg
([example](https://github.com/DataDog/datadog-agent/blob/valeri.pliskin/statkeeper-string-interning/pkg/network/encoding/encoding_test.go#L803))

Comment (hmahmood) [reply pkg/network/protocols/http/stats.go:18]: Can you explain why you need to use the same interner object? Maybe you can use a different (local) one for tests?
